### ISSUE TITLE
Unexpected WRN_NullReferenceReceiver warning around unconstrained type parameter

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#29983]
```
        [Fact]
        public void UnconstrainedTypeParameter_TypeInferenceThroughCall()
        {
            var source =
@"
class C
{
    static T Copy<T>(T t) => t;
    static void CopyOut<T>(T t1, out T t2) => t2 = t1;
    static void CopyOutInherit<T1, T2>(T1 t1, out T2 t2) where T1 : T2 => t2 = t1;
    static void M<U>(U u)
    {
        var x1 = Copy(u);
        x1.ToString(); // 1

        CopyOut(u, out var x2);
        x2.ToString(); // 2

        CopyOut(u, out U x3);
        x3.ToString(); // 3

        if (u == null) throw null;

        var x4 = Copy(u);
        x4.ToString();

        CopyOut(u, out var x5);
        x5.ToString();

        CopyOut(u, out U x6);
        x6.ToString();

        CopyOutInherit(u, out var x7);
        x7.ToString();
    }
}
";

            var comp = CreateCompilation(new[] { source, NonNullTypesAttributesDefinition, NonNullTypesTrue });
            // PROTOTYPE(NullableReferenceTypes): Should not report warning for `x6.ToString()`.
            comp.VerifyDiagnostics(
                // (29,9): error CS0411: The type arguments for method 'C.CopyOutInherit<T1, T2>(T1, out T2)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
                //         CopyOutInherit(u, out var x7);
                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "CopyOutInherit").WithArguments("C.CopyOutInherit<T1, T2>(T1, out T2)").WithLocation(29, 9),
                // (10,9): warning CS8602: Possible dereference of a null reference.
                //         x1.ToString(); // 1
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x1").WithLocation(10, 9),
                // (13,9): warning CS8602: Possible dereference of a null reference.
                //         x2.ToString(); // 2
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x2").WithLocation(13, 9),
                // (16,9): warning CS8602: Possible dereference of a null reference.
                //         x3.ToString(); // 3
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x3").WithLocation(16, 9),
                // (27,9): warning CS8602: Possible dereference of a null reference.
                //         x6.ToString();
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x6").WithLocation(27, 9));
        }
```

**Comment by jcouv** [comment]
I think this is by design now:we infer speakable types, so the fact that `u` is non-null is lost, so `x6` has a result type `U` which may be nullable.

**Comment by AlekseyTs** [comment]
@jcouv Did you mean to close the issue?

**Comment by jcouv** [comment]
Yes, thanks
